atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -367,6 +369,12 @@ public Optional<CassandraClientPoolingContainer> getRandomGoodHostForPredicate(
         return randomLivingHost.map(pools::get);
     }
 
+    public List<PoolingContainer<CassandraClient>> getAllHostsUnlessBlacklisted() {
+        Collection<CassandraServer> hosts =
+                blacklist.filterBlacklistedHostsFrom(getPools().keySet());
+        return hosts.stream().map(getPools()::get).collect(Collectors.toList());
```

Comment:
Do we need to consider the potential race conditions where current pools may be updated in the background? Should we operate over a consistent snapshot of `ImmutableMap.copyOf(currentPools)`? Can the `getPools()::get` return null and cause downstream NPE? should we filter out null results?

```suggestion
        ImmutableMap<CassandraServer, CassandraClientPoolingContainer> pools = ImmutableMap.copyOf(getPools());
        return blacklist.filterBlacklistedHostsFrom(pools.keySet()).stream()
                .map(pools::get)
                .filter(Objects::nonNull)
                .collect(ImmutableList.toImmutableList());
```